### PR TITLE
Standalone Lite: Precision Plugins

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds the precision plugins. These are the new, trimmed down interfaces. APIs that were Lightning specific have been removed. 

- Later on, PL will refactor their implementations to depend on these ones.
- New unit tests were added. Some existing ones were copied

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @carmocca @justusschock @awaelchli @akihironitta @rohitgr7

### CODE REVIEW

File: src/lightning_lite/plugins/__init__.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+from pytorch_lightning.plugins.precision.tpu_bf16 import TPUBf16PrecisionPlugin
+
+__all__ = [
+    "DeepSpeedPrecisionPlugin",
+    "DoublePrecisionPlugin",
+    "MixedPrecisionPlugin",
+    "NativeMixedPrecisionPlugin",
+    "PrecisionPlugin",
+    "TPUPrecisionPlugin",
+    "TPUBf16PrecisionPlugin",
```
Comment (awaelchli) [inline src/lightning_lite/plugins/__init__.py:16]: Reviewer vote:

👍 Remove the Plugin postfix from the names
👎 Keep as is

File: src/lightning_lite/plugins/precision/deepspeed.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    def optimizer_step(
+        self,
+        optimizer: Optimizer,
+        *args: Any,
+        model: Optional["deepspeed.DeepSpeedEngine"] = None,
+        **kwargs: Any,
+    ) -> Any:
+        if isinstance(optimizer, LBFGS):
+            raise TypeError("DeepSpeed and the LBFGS optimizer are not compatible.")
+        if model is None:
```
Comment (carmocca) [inline src/lightning_lite/plugins/precision/deepspeed.py:83]: Remind me, when would it be `None`?

File: src/lightning_lite/plugins/precision/native_amp.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        super().backward(tensor, model, *args, **kwargs)
+
+    def optimizer_step(
+        self,
+        optimizer: Optimizer,
+        *args: Any,
+        model: Optional[Module] = None,
+        **kwargs: Any,
+    ) -> Any:
+
```
Comment (carmocca) [inline src/lightning_lite/plugins/precision/native_amp.py:74]: nit:

```suggestion
```

File: src/lightning_lite/plugins/precision/precision.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    def forward_context(self) -> Generator[None, None, None]:
+        """A contextmanager for managing model forward/training_step/evaluation_step/predict_step."""
+        yield
+
+    def pre_backward(self, tensor: Tensor, module: Optional[Module]) -> None:
+        """Run before precision plugin executes backward.
+
+        Args:
+            tensor:
+            module:
```
Comment (carmocca) [inline src/lightning_lite/plugins/precision/precision.py:50]: Is this empty on purpose? I would fill it out or remove it. Other instances below

File: src/lightning_lite/plugins/precision/precision.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        """
+
+    def backward(self, tensor: Tensor, model: Optional[Module], *args: Any, **kwargs: Any) -> None:
+        """Performs the actual backpropagation."""
+        tensor.backward(*args, **kwargs)
+
+    def post_backward(self, tensor: Tensor, module: Optional[Module]) -> None:
```
Comment (carmocca) [inline src/lightning_lite/plugins/precision/precision.py:57]: The impl in PL does a `.detach()` here, should we too?

File: src/lightning_lite/plugins/precision/tpu.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+from typing import Any, Optional
+
+from torch.nn import Module
+from torch.optim import Optimizer
+
+from lightning_lite.plugins.precision.precision import PrecisionPlugin
+from lightning_lite.utilities.imports import _XLA_AVAILABLE
+
+if _XLA_AVAILABLE:
+    import torch_xla.core.xla_model as xm
```
Comment (carmocca) [inline src/lightning_lite/plugins/precision/tpu.py:23]: (I'm working on this in #14550) Can you import `torch_xla` locally in the method that uses it? This would be part of https://github.com/Lightning-AI/lightning/issues/12786

File: src/lightning_lite/plugins/precision/tpu.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+    def optimizer_step(
+        self,
+        optimizer: Optimizer,
+        *args: Any,
+        model: Optional[Module] = None,
+        **kwargs: Any,
+    ) -> Any:
+        if args:
+            raise ValueError("Positional arguments to the optimizer step need to be passed as named arguments.")
```
Comment (carmocca) [inline src/lightning_lite/plugins/precision/tpu.py:37]: Why this?

Also, there's a native way to do this:

```python
def optimizer_step(self, optimizer, *, model, **kwargs): ...
```

https://python-3-for-scientists.readthedocs.io/en/latest/python3_advanced.html

File: tests/tests_lite/plugins/precision/test_double.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+
+from lightning_lite.plugins.precision.double import DoublePrecisionPlugin
+
+
+def test_double_precision_forward_context():
```
Comment (carmocca) [inline tests/tests_lite/plugins/precision/test_double.py:20]: (meta question, can resolve): How are you tackling tests?

Are these moved from PL? If yes, will you remove them later?
If no: will they be duplicated?
I wonder what will be the coverage overlap between PL and Lite once this is finished.

**Review by carmocca — APPROVED** [review]

File: src/lightning_lite/plugins/__init__.py (hunk 1 above)
Comment (carmocca) [reply src/lightning_lite/plugins/__init__.py:16]: We chose the suffix because of the `plugins=` argument in the Trainer and because there was a shared Plugin interface.

Neither is true for Lite, so :+1:

File: tests/tests_lite/plugins/precision/test_double.py (hunk 8 above)
Comment (awaelchli) [reply tests/tests_lite/plugins/precision/test_double.py:20]: > Are these moved from PL? If yes, will you remove them later?

Some of them are, some of them are newly written by me. 


My workflow is:

1. Copy the original test file to the lite test folder. 
2. Delete all tests that require Trainer or similar PL interfaces, keep all unit tests that just test the component
3. Write new  unit tests that were not there in PL, to cover the important features and APIs of the component

After Lite has completely moved and all components have been integrated, a follow up workflow will be:

1. Rewrite integration tests that use the Trainer to Lite
2. Delete tests in PL that cover functionality completely implemented in Lite.

Over the next few weeks, we will see a very strong increase in tests in Lite with a small decrease in tests on PL. During the transition, lots of tests will be duplicated. When we remove redundant tests from PL, we want to make sure coverage does not decrease.

File: src/lightning_lite/plugins/precision/native_amp.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        if self.scaler is None:
+            # skip scaler logic, as bfloat16 does not require scaler
+            return super().optimizer_step(optimizer, *args, model=model, **kwargs)
+        if isinstance(optimizer, LBFGS):
+            raise TypeError("Native AMP and the LBFGS optimizer are not compatible.")
+
+        self.scaler.unscale_(optimizer)
+
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning_lite/plugins/precision/native_amp.py:82]: Yes you are right. It is here because it was in the code-path for non-PL execution, but it is optional. 

From amp docs:

> If [unscale_()](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler.unscale_) is not called explicitly, gradients will be unscaled automatically during [step()](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler.step).

File: tests/tests_lite/plugins/precision/test_double.py (hunk 8 above)
Comment (carmocca) [reply tests/tests_lite/plugins/precision/test_double.py:20]: > Delete all tests that require Trainer or similar PL interfaces, keep all unit tests that just test the component

Since we often use the "train-model-to-test" pattern by putting assertions inside a hook, this most tests will get deleted, making 

> Rewrite integration tests that use the Trainer to Lite

a big task down the line.

Thanks for the description!

File: src/lightning_lite/plugins/precision/tpu.py (hunk 7 above)
Comment (carmocca) [reply src/lightning_lite/plugins/precision/tpu.py:37]: Why not add the `, *,` pattern to this signature only? Or does that break mypy?

File: src/lightning_lite/plugins/precision/precision.py (hunk 5 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/precision/precision.py:57]: I'm mainly unhappy with the fact that we need to return the loss from the call here. That makes it inconsistent with the pre_backward. On top of that, the user would detach the loss themselves in their Lite code. It seems unnecessary to me.

If we add it, it would only exist for PL, but right now it is unclear how we will be able to integrate the Lite precision interface completely (and cleanly) with PL given the mess with apex.

File: src/lightning_lite/plugins/precision/tpu.py (hunk 7 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/precision/tpu.py:37]: Hmm, with this signature:
```

    def optimizer_step(
        self,
        optimizer: Optimizer,
        *,
        model: Optional[Module] = None,
        **kwargs: Any,
    ) -> Any:
```
I get
```
src/lightning_lite/plugins/precision/tpu.py:25: error: Signature of "optimizer_step" incompatible with supertype "PrecisionPlugin"  [override]
```

Comment (carmocca) [reply src/lightning_lite/plugins/precision/tpu.py:33]: Then IMO restrict the base class.

File: src/lightning_lite/plugins/precision/precision.py (hunk 5 above)
Comment (carmocca) [reply src/lightning_lite/plugins/precision/precision.py:62]: > right now it is unclear how we will be able to integrate the Lite precision interface completely (and cleanly) with PL given the mess with apex.

I don't get what would be the problem, but considering #14416, I would not give too much thought into having a nice clean apex plugin

**Review by tchaton — APPROVED** [review]
LGTM !

File: src/lightning_lite/plugins/precision/double.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from contextlib import contextmanager
+from typing import Generator
+
+import torch
+from torch import FloatTensor
```
Comment (justusschock) [inline src/lightning_lite/plugins/precision/double.py:18]:
```suggestion
```

File: src/lightning_lite/plugins/__init__.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/__init__.py]: We got votes from code owners Justus, Thomas, Carlos and I vote for this as well. So far no concerns so I'm going to push a commit with this change.

File: src/lightning_lite/plugins/precision/precision.py (hunk 5 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/precision/precision.py:54]: Resolving, open to change this once it becomes needed in Lite
